samza-kafka/src/main/java/org/apache/samza/system/kafka/KafkaSystemAdmin.java

```diff
@@ -0,0 +1,680 @@
+  private final Map<String, ChangelogInfo> changelogTopicMetaInformation;
+
+  // Kafka properties for intermediate topics creation
+  private final Map<String, Properties> intermediateStreamProperties;
+
+  // adminClient is required for deleteBeforeMessages operation
+  private final AdminClient adminClient;
+
+  private final boolean deleteCommittedMessages;
+
+  // used in TestRepartitionJoinWindowApp
+  @VisibleForTesting
+  public static volatile boolean deleteMessageCalled = false;
+
+  private final AtomicBoolean stopped = new AtomicBoolean(false);
+
+  @Override
+  public void start() {
+    // Plese note. There is slight inconsistency in the use of this class.
+    // Some of the functinality of this class may actually be used BEFORE start() is called.
+    // The SamzaContainer gets metadata (using this class) in SamzaContainer.apply, but this "start" actually gets called in SamzaContainer.run.
+
+    // Throw exception if start is called after stop
+    if (stopped.get()) {
+      throw new SamzaException("SamzaKafkaAdmin.start() is called after stop()");
+    }
+  }
+
+  @Override
+  public void stop() {
+    if (stopped.compareAndSet(false, true)) {
+      try {
+        metadataConsumer.close();
+      } catch (Exception e) {
+        LOG.warn("metadataConsumer.close for system " + systemName  + " failed with exception.", e);
+      }
+      try {
+        adminClient.close();
+      } catch (Exception e) {
+        LOG.warn("adminClient.close for system " + systemName  + " failed with exception.", e);
+      }
+    }
+  }
+
+  @VisibleForTesting
+  KafkaSystemAdmin(String systemName, Supplier<Consumer<K, V>> metadataConsumerSupplier,
+      Supplier<ZkUtils> connectZk, Supplier<AdminClient> connectAdminClient,
+      Map<String, ChangelogInfo> changelogTopicMetaInformation, Map<String, Properties> intermediateStreamProperties,
+      Properties coordinatorStreamProperties, int coordinatorStreamReplicationFactor, boolean deleteCommittedMessages) {
+    this.systemName = systemName;
+    this.metadataConsumer = metadataConsumerSupplier.get();
+    this.adminClient = connectAdminClient.get();
+    this.changelogTopicMetaInformation = changelogTopicMetaInformation;
+    this.intermediateStreamProperties = intermediateStreamProperties;
+    this.coordinatorStreamProperties = coordinatorStreamProperties;
+    this.coordinatorStreamReplicationFactor = coordinatorStreamReplicationFactor;
+    this.connectZk = connectZk;
+    this.connectAdminClient = connectAdminClient;
+    this.deleteCommittedMessages = deleteCommittedMessages;
+  }
+
+  /**
+   * Helper method to create KafkaSystemAdmin
+   * @param systemName system name
+   * @param config application config
+   * @param idPrefix will be used to generate clientId for the metadata consumer
+   * @return new KafkaSystemAdmin object
+   */
+  public static KafkaSystemAdmin getKafkaSystemAdmin(final String systemName, final Config config,
```

Comment:
Can we update the signature to return KafkaSystemAdmin<byte[], byte[]> to get rid of unchecked warnings?